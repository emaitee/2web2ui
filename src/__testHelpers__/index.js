export { default as delay } from './delay';
export { default as mockStore, createMockStore } from './mockStore';
export { default as renderRowData } from './renderRowData';
