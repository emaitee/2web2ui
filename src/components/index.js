export { default as ApiErrorBanner } from './apiErrorBanner/ApiErrorBanner';
export { default as CopyField } from './copyField/CopyField';
export { default as FilterDropdown } from './filterDropdown/FilterDropdown';
export { default as GlobalAlertWrapper } from './globalAlert/GlobalAlertWrapper';
export { default as LabelledValue } from './labelledValue/LabelledValue';
export { default as Layout } from './layout/Layout';
export { default as Modal } from './modals/Modal';
export { default as Navigation } from './navigation/Navigation';
export { default as PlanPicker } from './planPicker/PlanPicker';
export { default as SparkPost } from './sparkPost/SparkPost';
export { default as UsageReport } from './usageReport/UsageReport';
export { default as ApiKeySuccessBanner } from './apiKeyBanner/ApiKeyBanner';
export { default as Empty } from './empty/Empty';
export { default as PanelLoading } from './panelLoading/PanelLoading';
export { default as RouteWatch } from './routeWatch/RouteWatch';
export { default as LongTextContainer } from './longTextContainer/LongTextContainer';

export * from './collection';
export * from './formatters';
export * from './loading/Loading';
export * from './modals';
export * from './reduxFormWrappers';
