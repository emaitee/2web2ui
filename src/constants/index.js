export const DEFAULT_REDIRECT_ROUTE = '/landing';
