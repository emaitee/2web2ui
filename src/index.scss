@import '~@sparkpost/matchbox/styles.scss'; // Global Compiled
@import '~@sparkpost/matchbox/src/styles/config.scss';

// All components use REMs for sizing
// Adjusting the root font size here will scale the UI proportionally

html {
  background: color(gray, 9);
  font-size: rem(16);
  @media screen and (min-width: breakpoint(medium)) { font-size: rem(18); }
  @media screen and (min-width: breakpoint(larger)) { font-size: rem(20); }
}

::-webkit-input-placeholder { /* Chrome/Opera/Safari */
  color: #aaa;
}
::-moz-placeholder { /* Firefox 19+ */
  color: #aaa;
}
:-ms-input-placeholder { /* IE 10+ */
  color: #aaa;
}
:-moz-placeholder { /* Firefox 18- */
  color: #aaa;
}

// Overrides Critical CSS. Prevents FOUT.
#critical.ready {
  opacity: 0 !important;
  pointer-events: none !important;
}
