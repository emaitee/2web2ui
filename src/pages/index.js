export { default as AuthPage } from './auth/AuthPage';
export { default as billing } from './billing';
export { default as DashboardPage } from './dashboard/DashboardPage';
export { default as ProfilePage } from './profile/ProfilePage';
export { default as apiKeys } from './api-keys';
export { default as RegisterPage } from './register/RegisterPage';
export { default as reports } from './reports';
export { default as recipientLists } from './recipientLists';
export { default as sendingDomains } from './sendingDomains';
export { default as suppressions } from './suppressions';
export { default as subaccounts } from './subaccounts';
export { default as templates } from './templates';
export { default as users } from './users';
export { default as trackingDomains } from './trackingDomains';
export { default as webhooks } from './webhooks';
export { default as ipPools } from './ipPools';
export { default as SSOPage } from './sso/SSOPage';
export { default as ComingSoonPage } from './comingSoon/ComingSoonPage';
export { default as PageNotFound } from './pageNotFound/PageNotFound';
export { default as DefaultRedirect } from './defaultRedirect/DefaultRedirect';
