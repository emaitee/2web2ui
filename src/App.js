import React from 'react';

// Components
import { ProtectedRoute, AuthenticationGate, GlobalAlert } from 'src/components';

// Pages
import {
  AuthPage,
  billing,
  DashboardPage,
  ProfilePage,
  apiKeys,
  reports,
  templates,
  webhooks
} from './pages';

import {
  BrowserRouter as Router,
  Route,
  Redirect
} from 'react-router-dom';

const ForgotPassword = () => <h1>Forgot Password</h1>;

export default () => (
  <Router>
    <div>
      <AuthenticationGate />

      <Route exact path='/' render={() => <Redirect to='/auth' />} />
      <Route path='/auth' component={AuthPage} />
      <Route path='/forgot-password' component={ForgotPassword} />

      <ProtectedRoute path='/dashboard' component={DashboardPage} />
      <Route exact path='/reports' render={() => <Redirect to='/reports/summary' />} />
      <ProtectedRoute path='/reports/summary' component={reports.SummaryPage} />

      <ProtectedRoute exact path='/templates' component={templates.ListPage} />
      <ProtectedRoute exact path='/templates/create/' component={templates.CreatePage} />
      <ProtectedRoute exact path='/templates/edit/:id' component={templates.EditPage} />
      <ProtectedRoute exact path='/templates/edit/:id/published' component={templates.PublishedPage} />
      {/* <ProtectedRoute exact path='/templates/edit/:id/preview' component={TemplatesEditPage} /> */}

      <ProtectedRoute exact path='/webhooks' component={webhooks.ListPage}/>
      <ProtectedRoute exact path='/webhooks/create' component={webhooks.CreatePage}/>
      <ProtectedRoute path='/webhooks/details/:id' component={webhooks.DetailsPage}/>

      <ProtectedRoute exact path='/account/api-keys' component={apiKeys.ListPage} />
      <ProtectedRoute exact path='/account/api-keys/create' component={apiKeys.CreatePage} />
      <ProtectedRoute path='/account/api-keys/details/:id' component={apiKeys.DetailsPage} />

      <ProtectedRoute exact path='/account/profile' component={ProfilePage} />

      <ProtectedRoute exact path='/account/billing' component={billing.OverviewPage}/>
      <ProtectedRoute exact path='/account/billing/change' component={billing.ChangePlanPage}/>

      <GlobalAlert />
    </div>
  </Router>
);
