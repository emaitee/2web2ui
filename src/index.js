import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import config from './config';
import { unregister } from './helpers/registerServiceWorker';
import ErrorTracker from './helpers/errorTracker';
import rootReducer from './reducers';

import './critical.scss';
import './index.scss';
import App from './App';

const debug = () => (next) => (action) => {
  console.log(`=== ACTION: ${action.type} ===`); // eslint-disable-line no-console
  console.log(action); // eslint-disable-line no-console
  next(action);
};

// necessary for redux devtools in development mode only
const composeEnhancers = process.env.NODE_ENV !== 'production' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose; // eslint-disable-line no-mixed-operators
const store = createStore(
  rootReducer,
  composeEnhancers(
    applyMiddleware(thunk),
    applyMiddleware(debug),
    applyMiddleware(ErrorTracker.middleware)
  )
);

ErrorTracker.install(config, store);

render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById('root')
);
unregister(); // our bundle is currently too big to be added to SW cache, causing problems on every deploy

// Kill loading screen
document.getElementById('critical').className += ' ready';

/*  Google Analytics */
window.gtag('config', config.gaTag);

// appends tag manager script src to head
const gaScript = document.createElement('script');
gaScript.src = `https://www.googletagmanager.com/gtag/js?id=${config.gaTag}`;
gaScript.async = true;
document.head.appendChild(gaScript);
/*  Google Analytics End */

/**
 * Track unhandled promise rejects
 *
 * @param {PromiseRejectionEvent} event
 * @param {Error} event.reason
 */
window.onunhandledrejection = ({ reason }) => {
  ErrorTracker.report('onunhandledrejection', reason);
};
