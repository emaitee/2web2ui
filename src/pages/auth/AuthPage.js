import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import _ from 'lodash';
import { authenticate, ssoCheck, login } from 'src/actions/auth';
import { verifyAndLogin } from 'src/actions/tfa';
import { SparkPost } from 'src/components';
import { Panel, Error } from '@sparkpost/matchbox';
import { SubmissionError } from 'redux-form';
import selectAccessConditionState from 'src/selectors/accessConditionState';

import config from 'src/config';
import LoginForm from './components/LoginForm';
import TfaForm from './components/TfaForm';
import styles from './AuthPage.module.scss';

export class AuthPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ssoEnabled: config.sso.enabled
    };
  }

  ssoSignIn(username) {
    return this.props.ssoCheck(username).catch((err) => err);
  }

  regularSignIn(username, password, rememberMe) {
    return this.props.authenticate(username, password, rememberMe);
  }

  renderLoginError(errorDescription) {
    return (
      <Error error={errorDescription} />
    );
  }

  componentWillReceiveProps(nextProps) {
    const { ssoUser } = nextProps.auth;

    if (typeof ssoUser === 'undefined') { //anytime before action is dispatched
      return;
    }

    if (ssoUser) {
      window.location.assign(`${config.apiBase}/users/saml/login`);
    } else {
      this.setState({ ssoEnabled: false });
    }
  }

  loginSubmit = (values) => {
    const { username, password, rememberMe } = values;
    this.state.ssoEnabled ? this.ssoSignIn(username) : this.regularSignIn(username, password, rememberMe);
  }

  tfaSubmit = (values) => {
    const { code } = values;
    const { tfaEnabled, ...authData } = this.props.tfa;
    return this.props.verifyAndLogin({ authData, code }).catch((err) => {
      if (err.response.status === 400) {
        throw new SubmissionError({
          code: 'The code is invalid'
        });
      }
    });
  }

  render() {
    const { ready, location, currentUser } = this.props;
    const { errorDescription, loggedIn } = this.props.auth;
    const { tfaEnabled } = this.props.tfa;

    if (ready && loggedIn) {
      const redirectAfterLogin = _.get(location, 'state.redirectAfterLogin');
      const redirectTo = redirectAfterLogin || _.get(config, `entryPoints.${currentUser.access_level}`, config.splashPage);
      return <Redirect to={redirectTo} />;
    }

    return (
      <div>
        <div className={styles.LogoWrapper}>
          <a href="https://www.sparkpost.com" title="SparkPost">
            <SparkPost.Logo />
          </a>
        </div>

        <Panel sectioned accent title="Log In">
          { errorDescription && this.renderLoginError(errorDescription)}

          { tfaEnabled && <TfaForm onSubmit={this.tfaSubmit} /> }
          { !tfaEnabled && <LoginForm onSubmit={this.loginSubmit} ssoEnabled={this.state.ssoEnabled}/> }
        </Panel>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  const { auth, tfa } = state;
  return {
    auth,
    tfa,
    ...selectAccessConditionState(state)
  };
};

export default connect(mapStateToProps, { login, verifyAndLogin, authenticate, ssoCheck })(AuthPage);
