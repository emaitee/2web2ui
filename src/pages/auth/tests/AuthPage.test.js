import { shallow } from 'enzyme';
import React from 'react';

import { AuthPage } from '../AuthPage';
import LoginForm from '../components/LoginForm';
import TfaForm from '../components/TfaForm';
import { DEFAULT_REDIRECT_ROUTE } from 'src/constants';

const props = {
  auth: {
    loggedIn: false,
    loginPending: false
  },
  tfa: {
    tfaEnabled: false,
    username: 'bertha',
    token: 'tokey-token'
  },
  authenticate: jest.fn(),
  history: {
    push: jest.fn()
  }
};

let wrapper;

beforeEach(() => {
  props.ssoCheck = jest.fn(() => Promise.resolve());
  props.verifyAndLogin = jest.fn(() => Promise.resolve());
  wrapper = shallow(<AuthPage {...props} />);
});

afterEach(() => {
  jest.resetAllMocks();
});

it('renders correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('renders correctly when there is a login error', () => {
  wrapper.setProps({ auth: { errorDescription: 'uh oh!' }});
  expect(wrapper).toMatchSnapshot();
});

it('redirects when logged in', () => {
  wrapper.setProps({ auth: { loggedIn: true }});
  expect(props.history.push).toHaveBeenCalledWith(DEFAULT_REDIRECT_ROUTE);
});

it('should display tfa form when TFA is enabled', () => {
  wrapper.setProps({ tfa: { tfaEnabled: true }});
  expect(wrapper).toMatchSnapshot();
});

it('should verify tfa login on submit', () => {
  const authPage = wrapper.instance();
  const verifySpy = jest.spyOn(authPage.props, 'verifyAndLogin');
  authPage.tfaSubmit({ code: 'code' });
  expect(verifySpy).toHaveBeenCalled();
});

it('should throw a submission error when verifyAndLogin fails with 4xx error', () => {
  const authPage = wrapper.instance();
  authPage.props.verifyAndLogin.mockImplementation(() => Promise.reject({ response: { status: 400 }}));
  return authPage.tfaSubmit({ code: 'code' }).catch((err) => {
    expect(err.errors.code).toEqual('The code is invalid');
  });
});

it('should not throw an error when verifySpy fails with non 4xx error', async() => {
  const authPage = wrapper.instance();
  authPage.props.verifyAndLogin.mockImplementation(() => Promise.reject({ response: { status: 500 }}));
  await expect(authPage.tfaSubmit({ code: 'code' })).resolves.toBeUndefined();
});

it('should bind tfaSubmit to the submit handler of TfaForm', () => {
  wrapper.setProps({ tfa: { tfaEnabled: true }});
  const tfaForm = wrapper.find(TfaForm);
  expect(tfaForm.props().onSubmit).toBe(wrapper.instance().tfaSubmit);
});

it('should bind loginSubmit to the submit handler of LoginForm', () => {
  const loginForm = wrapper.find(LoginForm);
  expect(loginForm.props().onSubmit).toBe(wrapper.instance().loginSubmit);
});

it('should sso check on submit when sso is enabled', () => {
  const authPage = wrapper.instance();
  const ssoCheckSpy = jest.spyOn(authPage.props, 'ssoCheck');
  const authSpy = jest.spyOn(authPage.props, 'authenticate');
  wrapper.setState({ ssoEnabled: true });
  authPage.loginSubmit({ username: 'foo', password: 'pw', rememberMe: false });
  expect(ssoCheckSpy).toHaveBeenCalled();
  expect(authSpy).not.toHaveBeenCalled();
});

it('should authenticate on submit when sso is disabled', () => {
  const authPage = wrapper.instance();
  const ssoCheckSpy = jest.spyOn(authPage.props, 'ssoCheck');
  const authSpy = jest.spyOn(authPage.props, 'authenticate');
  wrapper.setState({ ssoEnabled: false });
  authPage.loginSubmit({ username: 'foo', password: 'pw', rememberMe: false });
  expect(ssoCheckSpy).not.toHaveBeenCalled();
  expect(authSpy).toHaveBeenCalled();
});

it('should redirect to sso if there is a sso user', () => {
  const redirectSpy = jest.spyOn(window.location, 'assign').mockImplementation();
  wrapper.setProps({ auth: { ssoUser: 'foo-bar' }});
  expect(redirectSpy).toHaveBeenCalledWith(expect.stringMatching(/saml\/login$/));
});

it('should set sso enabled if there ssoUser is null', () => {
  wrapper.setProps({ auth: { ssoUser: null }});
  expect(wrapper.state().ssoEnabled).toBeFalsy();
});
