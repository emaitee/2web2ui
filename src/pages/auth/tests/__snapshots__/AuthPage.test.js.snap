// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redirects when logged in and access control is ready 1`] = `
<Redirect
  push={false}
  to="/dashboard"
/>
`;

exports[`renders correctly 1`] = `
<div>
  <div>
    <a
      href="https://www.sparkpost.com"
      title="SparkPost"
    >
      <Logo />
    </a>
  </div>
  <Panel
    accent={true}
    sectioned={true}
    title="Log In"
  >
    <Connect(ReduxForm)
      onSubmit={[Function]}
      ssoEnabled={false}
    />
  </Panel>
</div>
`;

exports[`renders correctly when there is a login error 1`] = `
<div>
  <div>
    <a
      href="https://www.sparkpost.com"
      title="SparkPost"
    >
      <Logo />
    </a>
  </div>
  <Panel
    accent={true}
    sectioned={true}
    title="Log In"
  >
    <Error
      error="uh oh!"
    />
    <Connect(ReduxForm)
      onSubmit={[Function]}
      ssoEnabled={false}
    />
  </Panel>
</div>
`;

exports[`should display tfa form when TFA is enabled 1`] = `
<div>
  <div>
    <a
      href="https://www.sparkpost.com"
      title="SparkPost"
    >
      <Logo />
    </a>
  </div>
  <Panel
    accent={true}
    sectioned={true}
    title="Log In"
  >
    <Connect(ReduxForm)
      onSubmit={[Function]}
    />
  </Panel>
</div>
`;
